mars/services/web/core.py

```diff
@@ -0,0 +1,145 @@
+
+
+_web_address_local = threading.local()
+
+
+def set_web_address(web_address):
+    _web_address_local.web_address = web_address
+
+
+def get_web_address():
+    assert hasattr(_web_address_local, 'web_address'), 'Please set web_address first before get web address'
+    return _web_address_local.web_address
+
+
+class _HandlerException(Exception):
+
+    def __init__(self, exc_type, exc_value: BaseException, exc_traceback):
+        self.exc_type = exc_type
+        self.exc_value = exc_value
+        self.exc_traceback = exc_traceback
+
+
+class ServiceWebHandlerBase(MarsRequestHandler):
+    _api_instances = dict()
+    _api_cls = None
+
+    async def post(self, path, **_):
+        api_method_name = path
+        # params format: api_id[None], args, kwargs
+        api_id, args, kwargs = deserialize(self.request.body)
+        try:
+            if not api_id:
+                if hasattr(self, api_method_name):
+                    method = getattr(self, api_method_name)
+                else:
+                    method = getattr(self._api_cls, api_method_name)
+            else:
+                # call method of api instance
+                method = getattr(self._api_instances[api_id], api_method_name)
+            if kwargs:
+                # Some methods are decorated with @alru_cache, which doesn't support dict as part of cache key.
+                result = method(*args, **kwargs)
+            else:
+                result = method(*args)
+            if inspect.iscoroutine(result):
+                result = await result
+            self.write(serialize(result))
+        except Exception as e:
+            logger.info(f'Execute method {api_method_name} with {api_id, args, kwargs} failed, got exception {e}')
+            traceback.format_exc()
+            exc_type, exc_value, exc_traceback = sys.exc_info()
+            self.write(serialize(_HandlerException(exc_type, exc_value, exc_traceback)))
+
+    def __destroy_api__(self, api_id):
+        api = self._api_instances.pop(api_id, None)
+        print('Destroyed api %s', api)
+
+
+class ServiceWebAPIBase:
+    _service_name = None
+
+    def __init__(self, http_client: AsyncHTTPClient, api_id: int):
+        self._http_client = http_client
+        self._api_id = api_id
+
+    def __getattr__(self, method_name):
+        async def _func(*args, **kwargs):
+            return await self._post(self._http_client, method_name, *args, api_id=self._api_id, **kwargs)
+
+        return _func
+
+    def __del__(self):
+        try:
+            self._sync_post('__destroy_api__', api_id=self._api_id, req_config=dict(timeout=(0.1, 0.5)))
+        except:  # noqa: E722  # nosec  # pylint: disable=bare-except  # pragma: no cover
+            # server is closed
+            pass
+
+    @classmethod
+    async def _post(cls, http_client: AsyncHTTPClient, api_method_name: str,
+                    *args, api_id=None, req_config=None, **kwarg):
```

Comment:
It's wired to put the parameter`api_id` and `req_config` between `*args` and `**kwargs`.